### PR TITLE
refactor: modals to utilize the drawer or be on full-screen

### PR DESCRIPTION
**Pull request by sshanzel**
## Changes
- Convert all modals on mobile to be either on full screen or drawer.
- Initially, we thought of abstracting the drawer inside the Modal component, but as the development went on. I realized that we will have to pass another object to the Modal which is `drawerProps`. By doing that, when we render the Drawer inside of the Modal, a lot of things are happening that are unrelated to the Drawer itself, and by passing the drawer props, it makes no sense to not just render the Drawer on demand inside the file rather than be "abstracted".
- There are only a few Modals that require rendering the Drawer, and some of them require customizations of the content which you won't be able to do easily if the Drawer is abstracted.

**To view all the previews, I prepared a confluence page to make it all easier**. The changes here are based on discussion with Tomer.

https://dailydotdev.atlassian.net/wiki/spaces/HAN/pages/837517352/Modals+migration+to+either+full+screen+or+drawer

## Events

Did you introduce any new tracking events?
Don't forget to update the [Analytics Taxonomy sheet](https://docs.google.com/spreadsheets/d/18Lv7zXges9QfVX5VYL1a-Hyl0e1sQ3sLr0OK8YZWKXI/edit#gid=0)

Log the new/changed events below:

| Type   | event_name  | value |
|--------|-------------|-------|
| Change/New | event name  | extra: { ... } |

### On those affected packages:
- [ ] Have you done sanity checks in the webapp?
- [ ] Have you done sanity checks in the extension?
- [ ] Does this not break anything in companion?

### Did you test the modified components media queries?
- [ ] MobileL (420px)
- [ ] Tablet (656px)
- [ ] Laptop (1020px)

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

MI-129 #done

### CODE REVIEW

File: packages/shared/src/components/drawers/Drawer.tsx [hunk 1]
```diff
@@ -35,7 +40,7 @@ interface DrawerProps
   isFullScreen?: boolean;
   isClosing?: boolean;
   title?: string;
-  onClose(): void;
+  onClose(e: MouseEvent | KeyboardEvent | MessageEvent): void;
```
Comment (sshanzel) [inline packages/shared/src/components/drawers/Drawer.tsx:43]: You might see a lot of these changes, this is due to Drawer's requirement to fulfill outside click, it uses the native event handler.

File: packages/shared/src/components/modals/SubmitArticleModal.tsx [hunk 2]
```diff
@@ -202,18 +202,14 @@ export default function SubmitArticleModal({
...
         >
-          <div>
-            <p className="mb-2 text-theme-label-tertiary typo-callout">
+          <div className="flex flex-col gap-4">
+            <p className="text-theme-label-tertiary typo-callout">
               Found an interesting post? Do you want to share it with the
               community? Enter the post&apos;s URL / link below to add it to the
               feed.
             </p>
-            {!isEnabled && (
```
Comment (sshanzel) [inline packages/shared/src/components/modals/SubmitArticleModal.tsx:212]: The design doesn't have this anymore.

File: packages/shared/src/components/modals/SubmitArticleModal.tsx [hunk 3]
```diff
@@ -222,52 +218,50 @@ export default function SubmitArticleModal({
...
-              fieldType="tertiary"
-              name="articleUrl"
-              inputId="article_url"
-              label="Paste post url"
-              disabled={!isEnabled}
-              hint={urlHint}
-              valid={!urlHint}
-              valueChanged={onUrlChanged}
+          </div>
+          <p className="typo-callout">
```
Comment (sshanzel) [inline packages/shared/src/components/modals/SubmitArticleModal.tsx:222]: The diff looks crazy again, just a wrapper removal.

File: packages/shared/src/components/checklist/SquadChecklistCard.tsx [hunk 4]
```diff
@@ -5,19 +5,18 @@ import { Squad } from '../../graphql/sources';
...
+import { useViewSize, ViewSize } from '../../hooks';
+import { Drawer } from '../drawers';
 
 const SquadChecklistCard = ({ squad }: { squad: Squad }): ReactElement => {
   const { steps, isChecklistVisible, setChecklistVisible, isChecklistReady } =
     useSquadChecklist({
       squad,
     });
-  const { sidebarRendered } = useSidebarRendered();
+  const isMobile = useViewSize(ViewSize.MobileL);
```
Comment (kopancek) [inline packages/shared/src/components/checklist/SquadChecklistCard.tsx:19]: Instead of relying on javascript, can we render both and rely on CSS media queries and hide the element we don't want to see?

File: packages/shared/src/components/drawers/Drawer.tsx [hunk 5]
```diff
@@ -116,11 +121,16 @@ function BaseDrawer({
...
+            className={classNames(
+              'sticky -bottom-3 bg-background-default',
+              className?.close,
+            )}
+          >
             <Button
               variant={ButtonVariant.Float}
               className="mt-3 w-full"
-              onClick={onClose}
+              onClick={(e) => onClose(e.nativeEvent)}
```
Comment (kopancek) [inline packages/shared/src/components/drawers/Drawer.tsx:133]: Why was this change needed?

File: packages/shared/src/components/modals/PushNotificationModal.tsx [hunk 6]
```diff
@@ -1,13 +1,18 @@
...
 import { Modal, ModalProps } from './common/Modal';
 import { NotificationPromptSource } from '../../lib/analytics';
 import { usePushNotificationMutation } from '../../hooks/notifications';
+import { useViewSize, ViewSize } from '../../hooks';
+import { Drawer } from '../drawers';
+import ConditionalWrapper from '../ConditionalWrapper';
 
 function PushNotificationModal(modalProps: ModalProps): ReactElement {
   const { onRequestClose } = modalProps;
+  const isMobile = useViewSize(ViewSize.MobileL);
```
Comment (kopancek) [inline packages/shared/src/components/modals/PushNotificationModal.tsx:15]: Same as above, do we need to use the hook to display the drawer instead of CSS media query directly?

File: packages/shared/src/components/modals/common/Modal.tsx [hunk 7]
```diff
@@ -101,16 +101,18 @@ export function Modal({
...
     modalKindToOverlayClassName[kind],
     overlayClassName,
   );
   const modalClassName = classNames(
-    'modal relative flex max-w-full flex-col items-center rounded-16 border border-theme-divider-secondary bg-accent-pepper-subtlest antialiased shadow-2 focus:outline-none',
+    'modal relative flex max-w-full flex-col items-center border border-theme-divider-secondary bg-accent-pepper-subtlest antialiased shadow-2 focus:outline-none',
+    'h-full tablet:h-auto tablet:rounded-16',
     modalKindToClassName[kind],
-    modalSizeToClassName[size],
+    isMobile ? 'w-full' : modalSizeToClassName[size],
```
Comment (kopancek) [inline packages/shared/src/components/modals/common/Modal.tsx:115]: We could do this with css media queries, no?

File: packages/shared/src/components/checklist/SquadChecklistCard.tsx (hunk 4 above)
Comment (DragosIorgulescu) [reply packages/shared/src/components/checklist/SquadChecklistCard.tsx:19]: The only potential drawback to that is if we trigger any events or things like that, if both are rendered, we would be duplicating them? maybe something to consider on a case by case basis

File: packages/shared/src/components/modals/common/Modal.tsx (hunk 7 above)
Comment (DragosIorgulescu) [reply packages/shared/src/components/modals/common/Modal.tsx:115]: in this case, I'm echoing @kopancek on using media queries

**Review by DragosIorgulescu — COMMENTED** [review]
Looking great overall! I just have one curiosity on a higher level:

Would it make sense to update the `Modal` component itself to behave as a drawer on mobile by default, and maybe have some prop to force it to a modal on mobile in the edge cases where we need it? 

It feels like in 90% of cases we have the same content, it's just wrapped in either `Modal` or `Drawer`, maybe there's an opportunity to make things a little DRY-er here. Curious about your thoughts.

File: packages/shared/src/components/drawers/Drawer.tsx (hunk 5 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/Drawer.tsx:133]: Previously, the type was not needing to pass an event.

File: packages/shared/src/components/modals/common/Modal.tsx (hunk 7 above)
Comment (sshanzel) [reply packages/shared/src/components/modals/common/Modal.tsx:115]: Right. Yeah, we can convert this.

**Comment by sshanzel** [comment]
> Maybe we could refactor that into a separate hook or component?

This is what I mentioned in the description, at first what we wanted was to make this condition be part of the `Modal` component and contain a `isDrawerOnMobile` prop. But then I realized there are some needed customizations that made the implementation "ugly" on some components.

File: packages/shared/src/components/checklist/SquadChecklistCard.tsx (hunk 4 above)
Comment (sshanzel) [reply packages/shared/src/components/checklist/SquadChecklistCard.tsx:19]: Yes, one thing that easily comes to mind is closing on outside click. Either one of them will think you are always clicking outside its elements, hence closing it. There's probably some workaround, but let's see what are the pros and cons on both ends.

**Comment by DragosIorgulescu** [comment]
> > Maybe we could refactor that into a separate hook or component?
> 
> This is what I mentioned in the description, at first what we wanted was to make this condition be part of the `Modal` component and contain a `isDrawerOnMobile` prop. But then I realized there are some needed customizations that made the implementation "ugly" on some components.

Ah, no problem then, ignore my comment 🙏  thank you for the clarification

**Comment by sshanzel** [comment]
> Ah, no problem then, ignore my comment 🙏 thank you for the clarification

All cool. I will push another commit doing that, then we can discuss what we would prefer better. Thank you! 🚀

File: packages/shared/src/components/checklist/SquadChecklistCard.tsx (hunk 4 above)
Comment (rebelchris) [reply packages/shared/src/components/checklist/SquadChecklistCard.tsx:19]: Good point Milan.
The idea is that we should use CSS media queries wherever possible as the default.

There some edge cases where we need it in JS evals, then we prefer the useViewSize hook.

File: packages/shared/src/components/modals/PushNotificationModal.tsx [hunk 8]
```diff
@@ -27,19 +28,37 @@ function PushNotificationModal(modalProps: ModalProps): ReactElement {
...
       kind={Modal.Kind.FlexibleTop}
       size={Modal.Size.Medium}
+      isDrawerOnMobile
+      drawerProps={{ className: { drawer: 'pb-4', close: 'mx-4' } }}
     >
       <Modal.Header />
-      <Modal.Body>
-        <Modal.Title>Enable Push Notifications</Modal.Title>
-        <Modal.Text className="text-center">
+      <img
```
Comment (sshanzel) [inline packages/shared/src/components/modals/PushNotificationModal.tsx:35]: This is one of the parts where I thought it didn't seem right. Where we will be rendering a lot of hidden elements and display based on screen size.

File: packages/shared/src/components/modals/common/ModalClose.tsx [hunk 9]
```diff
@@ -44,7 +45,9 @@ function ModalCloseComponent(
   }: ModalCloseProps,
   ref: Ref<HTMLButtonElement>,
 ): ReactElement {
-  if (!onClick) {
+  const { isDrawer } = useModalContext();
```
Comment (sshanzel) [inline packages/shared/src/components/modals/common/ModalClose.tsx:48]: Modal close button is not necessary in Drawers since it has its own close button.

File: packages/shared/src/components/modals/common/ModalFooter.tsx [hunk 10]
```diff
@@ -16,8 +16,8 @@ export function ModalFooter({
   justify = Justify.End,
   view,
 }: ModalFooterProps): ReactElement {
-  const { activeView } = useContext(ModalPropsContext);
-  if (view && view !== activeView) {
+  const { activeView, isDrawer } = useContext(ModalPropsContext);
```
Comment (sshanzel) [inline packages/shared/src/components/modals/common/ModalFooter.tsx:19]: Footer is not needed on Drawers.

File: packages/shared/src/components/modals/common/ModalHeader.tsx [hunk 11]
```diff
@@ -37,8 +37,14 @@ export function ModalHeader({
   title,
   showCloseButton = true,
 }: ModalHeaderProps): ReactElement {
-  const { activeView, onRequestClose, tabs } = useContext(ModalPropsContext);
+  const { activeView, onRequestClose, tabs, isDrawer } =
+    useContext(ModalPropsContext);
   const modalTitle = title ?? (tabs ? activeView : undefined);
+
+  if (isDrawer) {
```
Comment (sshanzel) [inline packages/shared/src/components/modals/common/ModalHeader.tsx:44]: Header is not needed on drawers (this component contains title and close button).

**Comment by sshanzel** [comment]
@dailydotdev/web-team let me know what you guys think about having the Drawer abstracted from the Modal. As mentioned before, the key difference is that we will be rendering elements that are sometimes not necessary to be displayed on different screen sizes, but of course, those elements are hidden.

The `isModal` is now at its minimum, which is at the major components only. Modal and InteractivePopup.

File: packages/shared/src/components/modals/PushNotificationModal.tsx (hunk 8 above)
Comment (capJavert) [reply packages/shared/src/components/modals/PushNotificationModal.tsx:35]: We can make a quick drawer wrapper component which will internally check Modal context and then render or not render depending on `isDrawer`. 

```js
const DrawerImg = (props) => {
    const modalCtx = useModalContext()

    if (modalCtx.isDrawer) {
        return null
    }

    return (
        <img {...props} />
    )
}
```

```js
<Modal>
    <DrawerImg />
</Modal>
```

File: packages/shared/src/components/checklist/SquadChecklistCard.tsx [hunk 12]
```diff
@@ -49,27 +45,17 @@ const SquadChecklistCard = ({ squad }: { squad: Squad }): ReactElement => {
...
-    );
-  }
-
   return (
     <InteractivePopup
+      isDrawerOnMobile
+      drawerProps={{ className: { drawer: 'pb-4' } }}
       position={InteractivePopupPosition.RightEnd}
-      className="rounded-none !bg-transparent"
+      className="flex w-full max-w-[21.5rem] justify-center rounded-none"
```
Comment (kopancek) [inline packages/shared/src/components/checklist/SquadChecklistCard.tsx:57]: Why do we need a custom max width here?

**Review by capJavert — COMMENTED** [review]
I like the direction, some questions/ideas.

File: packages/webapp/pages/posts/[id]/index.tsx [hunk 13]
```diff
@@ -222,17 +225,19 @@ const PostPage = ({ id, initialData }: Props): ReactElement => {
...
-              authGradientBg,
-              'sticky left-0 top-0 z-max w-full justify-center gap-2 border-b border-theme-color-cabbage px-4 py-2',
-            ),
-            button: 'flex-1 tablet:max-w-[9rem]',
-          }}
-        />
-      ),
+      customBanner: shouldShowAuthBanner &&
+        !isLaptop &&
+        (isTablet || !shouldShowLogin) && (
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/webapp/pages/posts/[id]/index.tsx:230]: When the popup (shouldShowLogin) is opened, this banner should not beseen.

File: packages/shared/src/components/modals/common/ModalFooter.tsx [hunk 14]
```diff
@@ -16,8 +16,8 @@ export function ModalFooter({
...
+  if ((view && view !== activeView) || isDrawer) {
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/components/modals/common/ModalFooter.tsx:20]: Personally, I prefer what you are suggesting. I used to do that way back (separate null returns), but from all the feedback, it was preferred to have the conditions merged. We should maybe define some guidelines for this so it won't have to be mentioned again.

**Review by kopancek — COMMENTED** [review]
Maybe I'm weird, but I like this approach much more and it seems to be easier to use than the previous one.

We could maybe do a better job with how we display the footer in those exceptional cases where we still show footer content in the drawer, e.g. by providing another prop to the footer itself.

File: packages/webapp/pages/posts/[id]/index.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (capJavert) [reply packages/webapp/pages/posts/[id]/index.tsx:230]: Aha, just tricky naming, ok.

File: packages/shared/src/components/drawers/Drawer.tsx [hunk 15]
```diff
@@ -14,6 +14,15 @@ import { useOutsideClick } from '../../hooks/utils/useOutsideClick';
...
 import { Button } from '../buttons/Button';
 
+export type PopupEventType =
+  | MouseEvent
+  | KeyboardEvent
+  | MessageEvent
+  | React.MouseEvent
+  | React.KeyboardEvent;
+
+export type PopupCloseEvent = (e: PopupEventType) => void;
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/components/drawers/Drawer.tsx:24]: Ah, good catch. Yes, that should be named callback for func.

File: packages/shared/src/components/modals/RanksModal/DevCardFooter.tsx [hunk 16]
```diff
@@ -17,7 +17,7 @@ export default function DevCardFooter({
 }: DevCardFooterProps): ReactElement {
   return (
     <div className="pt-2">
-      <Modal.Subtitle>Generate your DevCard</Modal.Subtitle>
+      <Modal.Subtitle className="mb-2">Generate your DevCard</Modal.Subtitle>
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/components/modals/RanksModal/DevCardFooter.tsx:20]: Yes, in the component where we declared this, we removed these margins. This is because it should be defined on the consuming end. Unless it is a padding, it should be from what was expelained.

File: packages/webapp/pages/squads/[handle]/index.tsx [hunk 17]
```diff
@@ -194,7 +194,6 @@ const SquadPage = ({
           )}
         />
         <SquadPageHeader squad={squad} members={squadMembers} />
-        <SquadChecklistCard squad={squad} />
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/webapp/pages/squads/[handle]/index.tsx:197]: Moved it somewhere we actually manage the state to show this or not.

File: packages/webapp/pages/posts/[id]/index.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/webapp/pages/posts/[id]/index.tsx:230]: Same here https://github.com/dailydotdev/apps/pull/2815#discussion_r1528329639

File: packages/shared/src/components/checklist/SquadChecklistCard.tsx (hunk 12 above)
Comment (sshanzel) [reply packages/shared/src/components/checklist/SquadChecklistCard.tsx:57]: We added the `w-full` - on laptop, this would unnecessarily consume the full page.

File: packages/shared/src/components/modals/common/Modal.tsx [hunk 18]
```diff
@@ -88,10 +89,15 @@ export function Modal({
...
   steps,
+  isDrawerOnMobile,
+  drawerProps,
+  shouldCloseOnOverlayClick,
   ...props
 }: ModalProps): ReactElement {
   const stepTitle = steps ? steps?.[0].key : undefined;
   const tabTitle = tabs ? modalTabTitle(tabs[0]) : undefined;
+  const isMobile = useViewSize(ViewSize.MobileL);
+  const isDrawerOpen = isDrawerOnMobile && isMobile;
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/components/modals/common/Modal.tsx:100]: I think it is more preferred to use `is` or `has` since `showDrawer` sounds like it can a function.

File: packages/shared/src/components/modals/report/ReportModal.tsx [hunk 19]
```diff
@@ -55,9 +57,12 @@ export function ReportModal({
...
           data-testid="report_comment"
         />
+        {isMobile ? footer : null}
       </Modal.Body>
-      <Modal.Footer justify={footer ? Justify.Between : Justify.End}>
-        {footer}
+      <Modal.Footer
+        justify={footer && !isMobile ? Justify.Between : Justify.End}
+      >
+        {isMobile ? null : footer}
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/components/modals/report/ReportModal.tsx:65]: That's intentional, the conditions are the opposite.

File: packages/shared/src/components/modals/common/ModalFooter.tsx (hunk 10 above)
Comment (sshanzel) [reply packages/shared/src/components/modals/common/ModalFooter.tsx:19]: It is highly likely that it won't happen. Since the footer has a top border and is at the bottom-most part. It would clash with the built-in "Close" button.

**Review by capJavert — APPROVED** [review]

File: packages/shared/src/components/checklist/ChecklistCard.spec.tsx [hunk 20]
```diff
@@ -67,35 +67,4 @@ describe('ChecklistCard component', () => {
     const progress = await screen.findAllByTestId('checklist-card-progress');
     expect(progress).toHaveLength(2);
   });
-
-  it('should not render close button if onRequestClose is provided', async () => {
```
Comment (sshanzel) [inline packages/shared/src/components/checklist/ChecklistCard.spec.tsx:71]: Unused now since the recent changes.

File: packages/shared/src/components/modals/common/Modal.tsx [hunk 21]
```diff
@@ -44,35 +47,33 @@ const modalKindAndSizeToOverlayClassName: Partial<
   Record<ModalKind, Partial<Record<ModalSize, string>>>
 > = {
   [ModalKind.FlexibleTop]: {
-    [ModalSize.Small]: 'pt-10 mobileL:pt-20',
-    [ModalSize.Medium]: 'pt-10 mobileL:pt-20',
+    [ModalSize.Small]: 'tablet:pt-20',
+    [ModalSize.Medium]: 'tablet:pt-20',
```
Comment (rebelchris) [inline packages/shared/src/components/modals/common/Modal.tsx:51]: Are these just generic modal changes in regards to mobileUX?

File: packages/shared/src/components/modals/common/ModalFooter.tsx (hunk 10 above)
Comment (rebelchris) [reply packages/shared/src/components/modals/common/ModalFooter.tsx:19]: Also wondered if we could just make it dynamic based on high level component or perhpas just not render modalFooter at all.
But might miss some design edge-cases?

File: packages/shared/src/components/modals/PushNotificationModal.tsx [hunk 22]
```diff
@@ -1,10 +1,43 @@
...
+  }
+
+  return (
+    <img
+      className="mx-auto my-14"
+      src={cloudinary.notifications.big}
+      alt="A sample browser notification"
+    />
+  );
+};
```
Comment (rebelchris) [inline packages/shared/src/components/modals/PushNotificationModal.tsx:40]: Wait this is exactly the same, but one has classes?
Can't we just uniform them? *(or is it a placement issue?)

**Review by rebelchris — COMMENTED** [review]
Looks a lot better, just wondering if there's no way we could do it reverse?

Have something like ModalOrDrawer

```
if (isDrawer) <Drawer />

<Modal />
```

Because technically they just 2 separate components right, that shouldn't really care about each other.
Just share the same content basically.

File: packages/shared/src/components/modals/PushNotificationModal.tsx (hunk 22 above)
Comment (sshanzel) [reply packages/shared/src/components/modals/PushNotificationModal.tsx:40]: I did before, but we render two elements at the same time. Then Ante suggested this approach to avoid the double elements. https://github.com/dailydotdev/apps/pull/2815#discussion_r1528325281

File: packages/shared/src/components/modals/common/Modal.tsx (hunk 21 above)
Comment (sshanzel) [reply packages/shared/src/components/modals/common/Modal.tsx:51]: Yes, all of our Modal is going to be either full screen or drawer on mobile. No more specific width sizes.

**Review by DragosIorgulescu — APPROVED** [review]
I don't have much to add other than what has already been discussed, great effort here 👌  to echo what has been said, if we want to further improve on this, we might want to add a tech debt item to avoid delaying the release of mobile UX.

Of course, at some point we will need to take care of these tech debt items 😅

**Review by nensidosari — APPROVED** [review]
Nothing much to add here on my end as well, especially after our agreement on [thread](https://dailydotdev.slack.com/archives/C0650M2KX8E/p1710839348107749?thread_ts=1710491202.765389&cid=C0650M2KX8E). 

All the discussions here super valid

**Comment by sshanzel** [comment]
Just working on switching the close button to arrow icon.

**Review by rebelchris — APPROVED** [review]

File: packages/shared/src/components/modals/common/ModalFooter.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/components/modals/common/ModalFooter.tsx:20]: @dailydotdev/web-team would love to get some thoughts here, just so if no one objects, we can probably start doing it.
